src/lib/Types/Core.hs

```diff
@@ -556,18 +572,26 @@ data AtomBinding (n::S) =
    deriving (Show, Generic)
 
 data TopFunBinding (n::S) =
-   UnspecializedTopFun Int (Atom n)    -- num specialization args, definition
- | SpecializedTopFun (SpecializationSpec n) -- Original (unspecialized) function, specialization args
- | SimpTopFun          (NaryLamExpr n)
- | FFITopFun           (ImpFunName n)
+   -- This is for functions marked `@noinline`, before we've seen their use
+   -- sites and discovered what arguments we need to specialize on.
+   AwaitingSpecializationArgsTopFun Int (Atom n)
+   -- Specification of a specialized function. We still need to simplify, lower,
+   -- and translate-to-Imp this function. When we do that we'll store the result
+   -- in the `impCache`. Or, if we're dealing with ix method specializations, we
+   -- won't go all the way to Imp and we'll store the result in
```

Comment:
This seems contradictory with the fact that you do extend the imp cache in `evalRequiredSpecializations`?

Reply:
I think it's correct? The `IxMethodSpecialization` path of `evalRequiredSpecializations` calls `lowerTopLevelFun` which doesn't go all the way to imp.